### PR TITLE
Introduce ResourceOptionsManager

### PR DESCRIPTION
**Pull request by julianrex**
PRs must be submitted under the terms of our Contributor License Agreement [CLA](https://github.com/mapbox/mapbox-maps-ios/blob/main/CONTRIBUTING.md#contributor-license-agreement).

Fixes: < Link to related issues that will be fixed by this pull request, if they exist >

## Pull request checklist:
 - [X] Briefly describe the changes in this PR.
 - [ ] ~Write tests for all new functionality. If tests were not written, please explain why.~ Updated
 - [X] Document any changes to public APIs.
 - [X] Apply changelog label ('breaking change', 'bug :beetle:', 'build', 'docs', 'feature :green_apple:', 'performance :zap:', 'testing :100:') or use the label 'skip changelog'
 - [ ] ~Add an entry inside this element for inclusion in the `mapbox-maps-ios` changelog: `<changelog></changelog>`.~ Manual edit.
 - [ ] Update the migration guide, API Docs, Markdown files - Readme, Developing, etc

### Summary of changes

This PR replaces `CredentialsManager` with `ResourceOptionsManager` creating a struct based `ResourceOptions` that shadows the type from MapboxCoreMaps. The bulk of the other changes are related to this except for the following:

- Update the cache path to live in `.mapbox`. This aligns with other frameworks.
- CI creates the `MapboxAccessToken` file - since it appears that the build pre-action script is not being run
- Updated min/max latitude projection test
- Re-enabled `testCacheManagerIsReleasedAndHandlerRetainingManagerIsCalled`
- Fix MapView retain issue with integration tests
- Fix `ResourceEvent` and associated test

### CODE REVIEW

File: Sources/MapboxMaps/Foundation/Extensions/Core/ResourceOptions.swift [hunk 1]
```diff
@@ -3,7 +3,50 @@ import Foundation
 
 // MARK: - ResourceOptions
 
-extension ResourceOptions {
+/// Options to configure access to a resource
+public struct ResourceOptions {
+
+    /// The access token to access the resource. This should be a valid, non-empty,
+    /// Mapbox access token
+    public var accessToken: String
```
Comment (julianrex) [inline Sources/MapboxMaps/Foundation/Extensions/Core/ResourceOptions.swift:11]: I'd like us to consider making access token a custom type, for example an enum. Not unlike an optional, but instead of a `.none` case, we could have a `.search` or `.default` case that would look for the token in the Info.plist etc. This way it's more explicit and we could deprecate it later if we ever choose to.

File: Sources/MapboxMaps/Foundation/Extensions/Core/ResourceOptions.swift [hunk 2]
```diff
@@ -41,90 +84,122 @@ extension ResourceOptions {
...
+        hasher.combine(assetPathURL)
         hasher.combine(cacheSize)
         hasher.combine(tileStore)
         hasher.combine(tileStoreUsageMode)
-        return hasher.finalize()
+    }
+}
+
+extension ResourceOptions: CustomStringConvertible, CustomDebugStringConvertible {
+    private func redactedAccessToken() -> String {
```
Comment (julianrex) [inline Sources/MapboxMaps/Foundation/Extensions/Core/ResourceOptions.swift:155]: If we do add a custom AccessToken type, then this description logic would move there too.

File: Sources/MapboxMaps/Foundation/ResourceOptionsManager.swift [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+        }
+        return defaultInstance!
+    }
+
+    /// :nodoc:
+    /// Convenience function to remove the default instance. Calling `default`
+    /// again will re-create the default instance.
+    internal static func destroyDefault() {
+        defaultInstance = nil
+    }
```
Comment (julianrex) [inline Sources/MapboxMaps/Foundation/ResourceOptionsManager.swift:45]: This is in preparation of being able to clean up TileStores completely - useful for tests.

File: Sources/MapboxMaps/Foundation/ResourceOptionsManager.swift [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    /// we will use appropriate defaults for the `ResourceOptions`, including
+    /// searching for an access token in the application's Info.plist.
+    ///
+    /// You can override the shared global access token, using
+    /// `ResourceOptionsManager.default`:
+    ///
+    ///     ```
+    ///     ResourceOptionsManager.default.update { resourceOptions in
+    ///         resourceOptions.accessToken = "overridden-access-token"
+    ///     }
```
Comment (julianrex) [inline Sources/MapboxMaps/Foundation/ResourceOptionsManager.swift:75]: Old comments will be updated.

File: Sources/MapboxMaps/Foundation/ResourceOptionsManager.swift [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+    ///     ```
+    ///
+    /// - Parameter accessToken: Valid access token or `nil`
+    public convenience init(accessToken: String? = nil) {
```
Comment (julianrex) [inline Sources/MapboxMaps/Foundation/ResourceOptionsManager.swift:79]: Also added a convenience that just takes an access token.

File: Sources/MapboxMaps/Foundation/ResourceOptionsManager.swift [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+        self.bundle = bundle
+        update(resourceOptions)
+    }
+
+    /// Reset the manager to a default `ResourceOptions` using the specified
+    /// access token.
+    ///
+    /// - Parameter accessToken: Valid access token or `nil`
+    internal func reset(resourceOptions: ResourceOptions? = nil) {
+        self.resourceOptions = resourceOptions ?? ResourceOptions(accessToken: defaultAccessToken())
```
Comment (julianrex) [inline Sources/MapboxMaps/Foundation/ResourceOptionsManager.swift:105]: The default access token logic should probably move to where the core ResourceOptions object is initialized from the struct.

File: Sources/MapboxMaps/Foundation/ResourceOptionsManager.swift [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        let token = resourceOptions.accessToken.isEmpty ?
+            defaultAccessToken() : resourceOptions.accessToken
+
+        _resourceOptions = resourceOptions
+        _resourceOptions.accessToken = token
+
+        let resolvedTileStore = resourceOptions.tileStore ?? TileStore.default
+        if resourceOptions.tileStoreUsageMode != .disabled {
+            resolvedTileStore.setAccessToken(token)
+        }
```
Comment (julianrex) [inline Sources/MapboxMaps/Foundation/ResourceOptionsManager.swift:119]: @pengdev this is what we were talking about earlier. ResourceOptions on its down no longer does this, only the `ResourceOptionsManager`.

File: Tests/MapboxMapsTests/Foundation/Projection/ProjectionTests.swift [hunk 8]
```diff
@@ -87,5 +87,14 @@ class ProjectionTests: XCTestCase {
...
         XCTAssertEqual(southPole2.latitude, -85.051, accuracy: 0.001)
+
+        // Check clamping
+        let northMax = CLLocationCoordinate2D(latitude: Projection.latitudeMax, longitude: 0)
+        let northMaxMercator = projectorType.project(northMax, zoomScale: zoomScale)
+        XCTAssertEqual(northMercator.y, northMaxMercator.y, accuracy: 0.000001)
+
+        let southMin = CLLocationCoordinate2D(latitude: Projection.latitudeMin, longitude: 0)
+        let southMinMercator = projectorType.project(southMin, zoomScale: zoomScale)
+        XCTAssertEqual(southMercator.y, southMinMercator.y, accuracy: 0.000001)
```
Comment (julianrex) [inline Tests/MapboxMapsTests/Foundation/Projection/ProjectionTests.swift:98]: Unrelated test - will likely leave in this PR for simplicity

File: Tests/MapboxMapsTests/MapView/Integration Tests/Map/DidIdleFailureIntegrationTest.swift [hunk 9]
```diff
@@ -33,7 +34,7 @@ internal struct ResourceEventResponse: Decodable {
     var noContent: Bool
     var notModified: Bool
     var mustRevalidate: Bool
-    var offlineData: Bool
+    var source: String
```
Comment (julianrex) [inline Tests/MapboxMapsTests/MapView/Integration Tests/Map/DidIdleFailureIntegrationTest.swift:37]: Unrelated test fix.

File: Sources/MapboxMaps/Foundation/Extensions/Core/CacheManager.swift [hunk 10]
```diff
@@ -1,6 +1,13 @@
 import Foundation
 
-extension CacheManager {
+extension MapboxCoreMaps.CacheManager {
```
Comment (neelmistry94) [inline Sources/MapboxMaps/Foundation/Extensions/Core/CacheManager.swift:3]: I know there was some convo around this, but does this need the full qualified name? Can we type alias it?

File: Sources/MapboxMaps/Foundation/Extensions/Core/ResourceOptions.swift (hunk 1 above)
Comment (nishant-karajgikar) [reply Sources/MapboxMaps/Foundation/Extensions/Core/ResourceOptions.swift:11]: I like this idea.. may be we could even make it `expressibleByString` in some way?

**Review by nishant-karajgikar — APPROVED** [review]

File: Sources/MapboxMaps/Foundation/Extensions/Core/MapSnapshotOptions.swift [hunk 11]
```diff
@@ -10,19 +10,21 @@ extension MapSnapshotOptions {
...
     ///   - resourceOptions: Resource fetching options to be used by the
     ///         snapshotter. Default uses the access token provided by
-    ///         `CredentialsManager.default`
+    ///         `ResourceOptionsManager.default`
     public convenience init(size: CGSize,
                             pixelRatio: CGFloat,
-                            glyphsRasterizationOptions: GlyphsRasterizationOptions? = GlyphsRasterizationOptions(fontFamilies: []),
-                            resourceOptions: ResourceOptions = ResourceOptions()) {
+                            glyphsRasterizationOptions: GlyphsRasterizationOptions? = GlyphsRasterizationOptions(),
+                            resourceOptions: ResourceOptions = ResourceOptionsManager.default.resourceOptions) {
```
Comment (pengdev) [inline Sources/MapboxMaps/Foundation/Extensions/Core/MapSnapshotOptions.swift:17]: MapSnapshotOptions is a gl-native generated class in Android, and we couldn't override the default value here...

File: Sources/MapboxMaps/Foundation/Extensions/Core/ResourceOptions.swift (hunk 1 above)
Comment (pengdev) [reply Sources/MapboxMaps/Foundation/Extensions/Core/ResourceOptions.swift:11]: cc @Chaoba 
on Android we can not overwrite the type in the ResourceOptions..

File: Sources/MapboxMaps/Foundation/Extensions/Core/ResourceOptions.swift [hunk 12]
```diff
@@ -41,90 +84,122 @@ extension ResourceOptions {
...
+                                                                in: .userDomainMask,
+                                                                appropriateFor: nil,
+                                                                create: true)
 
-    private static func cacheURLIncludingSubdirectory() throws -> URL? {
-        var cacheDirectoryURL = try FileManager.default.url(for: .applicationSupportDirectory,
-                                                            in: .userDomainMask,
-                                                            appropriateFor: nil,
-                                                            create: true)
+            cacheDirectoryURL = cacheDirectoryURL.appendingPathComponent(".mapbox")
```
Comment (pengdev) [inline Sources/MapboxMaps/Foundation/Extensions/Core/ResourceOptions.swift:110]: so you make this change in the same PR? probably need to include this change to the change log.

File: Sources/MapboxMaps/Foundation/MapView.swift [hunk 13]
```diff
@@ -102,7 +102,7 @@ open class MapView: UIView {
     /// - Parameters:
     ///   - frame: frame for the MapView.
     ///   - mapInitOptions: `MapInitOptions`; default uses
-    ///    `CredentialsManager.default` to retrieve a shared default access token.
+    ///    `ResourceOptionsManager.default` to retrieve a shared default access token.
```
Comment (pengdev) [inline Sources/MapboxMaps/Foundation/MapView.swift:105]:
```suggestion
    ///    `ResourceOptionsManager.default` to retrieve a shared default resource option, including the access token.
```

File: Sources/MapboxMaps/Foundation/ResourceOptionsManager.swift [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+/// `application(_:didFinishLaunchingWithOptions:)` rather than relying on the
+/// value in your application's Info.plist:
+///
+///     ```
+///     func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
+///         // Override point for customization after application launch.
+///         ResourceOptionsManager.default.resourceOptions.accessToken = "overridden-access-token"
+///         return true
+///     }
+///     ```
```
Comment (pengdev) [inline Sources/MapboxMaps/Foundation/ResourceOptionsManager.swift:19]: @Chaoba let's also align on the documentation

File: Sources/MapboxMaps/Foundation/ResourceOptionsManager.swift [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+    /// This shared instance is used by the default initializers
+    /// for `ResourceOptions` and `MapInitOptions`.
+    ///
+    /// The application's Info.plist will be searched for a valid access token
+    /// under the key `MBXAccessToken`
+    ///
+    /// A valid access token must be provided or found.
+    public static var `default`: ResourceOptionsManager {
+        if defaultInstance == nil {
+            defaultInstance = ResourceOptionsManager(accessToken: nil)
```
Comment (pengdev) [inline Sources/MapboxMaps/Foundation/ResourceOptionsManager.swift:33]: on iOS, do you throw exceptions when the accessToken is null?

File: Sources/MapboxMaps/Foundation/ResourceOptionsManager.swift (hunk 3 above)
Comment (pengdev) [reply Sources/MapboxMaps/Foundation/ResourceOptionsManager.swift:43]: we are considering to expose the same cc @Chaoba

File: Sources/MapboxMaps/Foundation/ResourceOptionsManager.swift [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+        }
+        // Check for a bundled file
+        else if let url = bundle.url(forResource: "MapboxAccessToken", withExtension: nil),
+                let tokenFromFile = try? String(contentsOf: url) {
+            token = tokenFromFile
+        }
+
+        if token.isEmpty {
+            Log.warning(forMessage: "Empty access token.", category: "ResourceOptions")
+        }
```
Comment (pengdev) [inline Sources/MapboxMaps/Foundation/ResourceOptionsManager.swift:143]: So you don't throw exceptions when the token is not found, and instead just show a warning message? What does it look like to users? will the MapView will show a black screen?

Comment (pengdev) [reply Sources/MapboxMaps/Foundation/ResourceOptionsManager.swift:143]: fyi, if the access token is not found anywhere, we will throw a runtime exception that crashes the app..

cc @Chaoba

File: Sources/MapboxMaps/Foundation/Extensions/Core/ResourceOptions.swift (hunk 12 above)
Comment (julianrex) [reply Sources/MapboxMaps/Foundation/Extensions/Core/ResourceOptions.swift:110]: Yep good call. Adding here for convenience :)

File: Sources/MapboxMaps/Foundation/ResourceOptionsManager.swift (hunk 15 above)
Comment (julianrex) [reply Sources/MapboxMaps/Foundation/ResourceOptionsManager.swift:33]: Not at the moment, does Android? Also, `nil` in this context means "search for the token".

File: Sources/MapboxMaps/Foundation/ResourceOptionsManager.swift (hunk 16 above)
Comment (julianrex) [reply Sources/MapboxMaps/Foundation/ResourceOptionsManager.swift:143]: We could do the same - though there's some trickiness around testing.

File: Sources/MapboxMaps/Foundation/ResourceOptionsManager.swift (hunk 7 above)
Comment (julianrex) [reply Sources/MapboxMaps/Foundation/ResourceOptionsManager.swift:115]: This has been removed in #429
